### ISSUE TITLE
Feature Request: Allow ES metadata field be available terms in alerts

### ISSUE THREAD

**Issue by romeotheriault** [Yelp/elastalert#354]
What I'm looking to do is, in the emails that elastalert sends out, build Kibana url's to the exact message (doc) in kibana. This requires that we have '_id' and '_index' available as terms we can use in the email/jira alerts. From looking through the code it seems that you're only using the terms in '_source'. 

Is there currently some way to access the values outside of _source? If not, what would be the cleanest way to get this added?

Thanks

**Comment by jraby** [comment]
I wanted the same thing, but in the end I figured that I didn't need the `_id` to do what I wanted.

I just had a quick read though the code again and it looks like only `hit['_source']` is being passed as a match, `remove_duplicate_events` drops everything else about each hits: https://github.com/Yelp/elastalert/blob/master/elastalert/elastalert.py#L343

Not too sure what would be a clean way to passes these fields along with the matches. It would be easy to set `_id`, `_type` and `_index`  in the `hit['_source']`, but that doesn't feel right (and that would overwrite the value of these fields in `_source` if you somehow had such fields.)

**Comment by Qmando** [comment]
I am not opposed to adding ``_type``, ``_id``, etc into _source. It's a bit of a hack, but even in Kibana, those fields along with sources fields are flattened into a table. In fact, I don't think it's possible to actually have an ``_id`` field in the document source and not have it be THE ``_id`` field.

**Comment by romeotheriault** [comment]
Thanks jraby, that's the info I needed. 

I patched my version of elastalert to include these lines:

            # Add _id, _index and _type into _source
            hit['_source']['_id'] = hit['_id']
            hit['_source']['_index'] = hit['_index']
            hit['_source']['_type'] = hit['_type']

right above this:
https://github.com/Yelp/elastalert/blob/master/elastalert/elastalert.py#L239

and it's working well.

Let me know if this is pull request worthy and I can send a pull request. Would be great to not have to patch elastalert going forward. 

Also, in regards to _id and other metadata fields in the source of a document. It seems that in ES 2.0+ they locked down on this so it's no longer possible, but pre-ES 2.0 there was a way to do it. See:

https://github.com/elastic/elasticsearch/pull/14003

**Comment by jraby** [comment]
Yeah it is a bit of a hack, but if a document cannot have these fields in `_source`, then I think it makes sense.

If we do this,  `_id` would be accessible via `event['_source']['_id']`, so maybe it would make sense to move this from `remove_duplicate_events`  to `process_hits`:
```
return [event['_source'] for event in data]
```

I was a bit surprised/puzzled to see that events data gets filtered in a function that removes duplicates.

What do you think @Qmando ?
